### DISCUSSION TITLE
How to check for the projectivized sentences with cycles?

### DISCUSSION THREAD

**Discussion by kanayer** [explosion/spaCy#10702] [Help: Coding & Implementations]
Hello! I am trying to train the Thai spaCy model on a custom dataset and when I run `spacy debug data -V` I get the following error message (in red): 

<img width="894" alt="Screen Shot 2022-04-25 at 4 54 36 PM" src="https://user-images.githubusercontent.com/22777876/165044882-5032f6bf-b41d-47fd-9cac-24e7a66e89a2.png">

I run the following code to check for the projectivized trees however, the code didn't find any. I was wondering if there is any other way to check for the projectivized sentences with cycles?

```import spacy
from spacy import displacy
from spacy.tokens import DocBin
from spacy.pipeline._parser_internals.nonproj import is_nonproj_tree

nlp = spacy.blank("xx")  
docbin = DocBin().from_disk("Thai/tagger_parser_ud/corpus/UD_Thai-MindAI/train.spacy")
for doc in docbin.get_docs(nlp.vocab):
    if is_nonproj_tree([t.head.i for t in doc]):
        print(doc)```

**Comment by adrianeboyd** [comment]
Hmm, that error shouldn't really happen just due to the projectivization. You might need to look into exactly what's going on in `debug data` here with this particular doc:

https://github.com/explosion/spaCy/blob/497a708c7130cda95fd08b678cb7adc109ebde0e/spacy/cli/debug_data.py#L735-L745

**Reply by kanayer** [reply]
I have run the code below however, there was no output. What can be the cause of the above-mentioned error message?

```
import spacy
nlp = spacy.blank("xx")
pipeline = nlp.pipe_names
factory_names = [nlp.get_pipe_meta(pipe).factory for pipe in nlp.pipe_names]

if "parser" in factory_names:
    aligned_heads, aligned_deps = eg.get_aligned_parse(projectivize=make_proj)
    data["deps"].update([x for x in aligned_deps if x is not None])
    for i, (dep, head) in enumerate(zip(aligned_deps, aligned_heads)):
        if head == i:
            data["roots"].update([dep])
            data["n_sents"] += 1
    if nonproj.is_nonproj_tree(aligned_heads):
        data["n_nonproj"] += 1
    if nonproj.contains_cycle(aligned_heads):
        data["n_cycles"] += 1
```

**Reply by adrianeboyd** [reply]
That code is only intended to work within that method/loop in `spacy debug data`. You can hack in some debugging to figure out which doc actually contains the cycle, or you can extract all the relevant parts of the code (the corpus loading and the compile gold method) into a separate script to debug in some other way.

**Reply by kanayer** [reply]
I'm sorry but could you please explain in more detail how can I hack the `debug_data.py` code to find which doc contains the cycle? I have cloned the spacy repo, and run the `debug_data.py` code with some test print messages, but it does not do anything. Could you please help me understand how to find the cyclic sentence?

**Reply by adrianeboyd** [reply]
You can try something like this in a separate script:

```python
import spacy
from spacy.training import Corpus
from spacy.pipeline._parser_internals import nonproj


corpus = Corpus("train.spacy")
nlp = spacy.blank("en")
examples = list(corpus(nlp))
for eg in examples:
    aligned_heads, aligned_deps = eg.get_aligned_parse()
    if nonproj.is_nonproj_tree(aligned_heads):
        print("non-projective:", eg.reference.text)
    if nonproj.contains_cycle(aligned_heads):
        print("contains cycle:", eg.reference.text)
```
